### PR TITLE
feat: 🎸 support ternary operator in structural directive

### PR DESCRIPTION
**Pull request by MikeDabrowski**
Closes: 41

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/ngneat/transloco/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features) (I think it was there already)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: #41 

## What is the new behavior?
Key Manager will parse the following and add it to json.
`<p>{{t(condition ? 'yin' : 'yang')}}</p>`
It won't work with multiply nested ternary operators.

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

## Other information

I'm not good in regexes so that for sure can be optimized. Also Im not sure if this is even right way to do this.

### CODE REVIEW

File: __tests__/read/1.html [hunk 1]
```diff
@@ -1,3 +1,8 @@
+<ng-container *transloco="let t; read: 'ternary.nested'">
+  <p>{{t('test')}}</p>
+  <p>{{t(condition ? 'yin' : 'yang')}}</p>
```
Comment (shaharkazaz) [inline __tests__/read/1.html:3]: Please add more permutations and see that your RegExp is handling them correctly.
For example: `condition ? "yin" : 'yang'` won't work

File: src/keysBuilder/getStructuralDirectiveBasedKeys.ts [hunk 2]
```diff
@@ -31,12 +31,20 @@ export function getStructuralDirectiveBasedKeys(
...
     let keySearch = keyRegex.exec(matchedStr);
+    let ternarySearch = ternaryRegex.exec(matchedStr);
 
     while (keySearch) {
       translationKeys.push(keySearch.groups.key);
       keySearch = keyRegex.exec(matchedStr);
     }
+
+    while (ternarySearch) {
+      translationKeys.push(ternarySearch.groups.keyA);
```
Comment (shaharkazaz) [inline src/keysBuilder/getStructuralDirectiveBasedKeys.ts:44]: keep it DRY:
```
const {keyA, keyB} = ternarySearch.groups;
translationKeys.push(keyA, keyB);
```

**Comment by shaharkazaz** [comment]
@MikeDabrowski Did you run the tests and checked that they are passing? 🙂 
It doesn't make sense that they do since you didn't update the `buildTranslationFiles.spec.ts` file, which runs the spec and checks the expected result vs. the actual result.

**Comment by MikeDabrowski** [comment]
No, sorry. Yesterday was a bad day. I might have pushed to the wrong branch. I will fix that tomorrow.

**Comment by shaharkazaz** [comment]
@MikeDabrowski It's all good, take your time. no rush on my side.
Can you also please align the keys in the test cases you added? (I'm mostly using numbers) just add the keys as the following numbers.
And add a ternary case that the manager shouldn't extract 👍

**Comment by MikeDabrowski** [comment]
@shaharkazaz 
I've made  changes but I got stuck on nested ternary. I think we should ban it but I have no idea how to do it on regex level. Any hint ? With  current regex I believe it will take first and last case.

**Comment by shaharkazaz** [comment]
@MikeDabrowski I agree we shouldn't support nested.
WDYM by banning it in a regex level? if I understand correctly your current regex is ignoring nested (like expected)

**Comment by MikeDabrowski** [comment]
@shaharkazaz 

For copypasta to regex101
`\bt\((.*?)\?[^:"']*("|')(?<keyA>[^'"?]*)\2(.*?)("|')(?<keyB>[^'"?]*)\5\)`
`<p>{{t(condition ? condition2 ? "10" : "11" : "12")}}</p>`

image


This is happening. 
Imo it would be better to change regex so it stops when it has second ? or something similar to prevent checks for nested ternary.

**Comment by shaharkazaz** [comment]
@MikeDabrowski I think your regex is a bit loose to the ternary pattern, they this:
`\bt\([^?}]+\?\s*("|')(?<keyA>[^'":]+)\1\s*:\s*("|')(?<keyB>[^'"]*)\3\s*\)`
This needs to be tested in a complex template just to make sure we don't have some edge cases. 😄

**Comment by MikeDabrowski** [comment]
image

The one you provided seems to be working though. WDYM exactly by complex template?

**Comment by shaharkazaz** [comment]
@MikeDabrowski I mean create a template that is combined with some components, regular data, and translations.
The test cases you added are pretty sterilized, look at the translations that are tested there, they are part of "a real template" and not just ternary one after the other 🙂 

Can you also update the README section with an example?

**Comment by shaharkazaz** [comment]
@MikeDabrowski any update on that?

**Comment by MikeDabrowski** [comment]
Sorry, I had no time. But I will come back to it as soon as I can.

**Comment by shaharkazaz** [comment]
@MikeDabrowski Any updates?  you just have the specs left

**Comment by MikeDabrowski** [comment]
I'm sorry. I remember about this but still can't find time. I will try this
sunday.

**Comment by MikeDabrowski** [comment]
I added few more bits deeper into the template but I don't quite feel how you want to test it in a more complicated template. Imo the sterile tests are sufficient because they test regex explicitly and the extraction from deeper templates is tested by in other cases.

File: src/regexs.ts [hunk 3]
```diff
@@ -10,6 +10,9 @@ export const regexs = {
   templateKey: varName => new RegExp(`\\b${varName}\\((?![^,)+]*\\+)('|")(?<key>[^)"']*?)\\1`, 'g'),
   directive: () => new RegExp(`\\stransloco\\s*=\\s*("|')(?<key>[^]+?)\\1`, 'g'),
   directiveTernary: () => new RegExp(`\\s\\[transloco\\]\\s*=\\s*("|')[^"'?]*\\?(?<key>[^]+?)\\1`, 'g'),
+  structuralDirectiveTernary: varName =>
+    // new RegExp(`\\b${varName}\\((.*?)\\?[^:"']*("|')(?<keyA>[^'"?]*)\\2(.*?)("|')(?<keyB>[^'"?]*)\\5\\)`, 'g'),
```
Comment (shaharkazaz) [inline src/regexs.ts:14]: Please remove the comment
